integration-tests/src/nearcore_utils.rs

```diff
@@ -0,0 +1,124 @@
+use crate::genesis_helpers::genesis_block;
+use actix::Addr;
+use near_chain::Block;
+use near_chain_configs::Genesis;
+use near_client::{BlockResponse, ClientActor};
+use near_network::tcp;
+use near_network::test_utils::convert_boot_nodes;
+use near_network::types::PeerInfo;
+use near_o11y::WithSpanContextExt;
+use near_primitives::block::Approval;
+use near_primitives::hash::CryptoHash;
+use near_primitives::merkle::PartialMerkleTree;
+use near_primitives::num_rational::{Ratio, Rational32};
+use near_primitives::types::validator_stake::ValidatorStake;
+use near_primitives::types::{BlockHeightDelta, EpochId};
+use near_primitives::validator_signer::ValidatorSigner;
+use near_primitives::version::PROTOCOL_VERSION;
+use nearcore::config::{GenesisExt, TESTING_INIT_STAKE};
+use nearcore::{load_test_config, NearConfig};
+
+// This assumes that there is no height skipped. Otherwise epoch hash calculation will be wrong.
+pub fn add_blocks(
+    mut blocks: Vec<Block>,
+    client: Addr<ClientActor>,
+    num: usize,
+    epoch_length: BlockHeightDelta,
+    signer: &dyn ValidatorSigner,
+) -> Vec<Block> {
+    let mut prev = &blocks[blocks.len() - 1];
+    let mut block_merkle_tree = PartialMerkleTree::default();
+    for block in blocks.iter() {
+        block_merkle_tree.insert(*block.hash());
+    }
+    for _ in 0..num {
+        let epoch_id = match prev.header().height() + 1 {
+            height if height <= epoch_length => EpochId::default(),
+            height => {
+                EpochId(*blocks[(((height - 1) / epoch_length - 1) * epoch_length) as usize].hash())
+            }
+        };
+        let next_epoch_id = EpochId(
+            *blocks[(((prev.header().height()) / epoch_length) * epoch_length) as usize].hash(),
```

Comment:
Extract the arithmetic expression to a variable

Reply:
I just moved this code and didn't want to mess with it, but yes, this is unreadable.
I hope I didn't break anything while refactoring.